### PR TITLE
CN0435 PLC system application

### PR DESCRIPTION
**Pull request by adrimbarean**
Add the code for the PLC/DCS system application. The project uses the CN0414 and CN0418 projects to make a dynamic system which acts as an industrial, input/output, MODBUS slave platform.

The first commit only adds already existing files to the project and thus does not require review.

### CODE REVIEW

**Comment by adrimbarean** [comment]
V0 -> V1:
 - correct bugs reported by travis

File: projects/ADuCM3029_demo_cn0435/src/cli.c [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+	int32_t ret;
+
+	if(!dev)
+		return -1;
+
+	ret = usr_uart_remove(dev->uart_device);
+	if(ret < 0)
+		return ret;
+
+	free(dev);
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/cli.c]: should we still free dev if uart_remove fails ?

File: projects/ADuCM3029_demo_cn0435/src/cli.c [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t cli_parse(struct cli_desc *dev)
+{
+	uint8_t c = 0, rdy = 1;
+	int8_t i;
+
+	usr_uart_read_nb(dev->uart_device, &c, 1, &rdy);
+
+	if(rdy == 1) {
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/cli.c]: if (rdy)

File: projects/ADuCM3029_demo_cn0435/src/cli.c [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+/**
+ * Implements CLI feedback.
+ *
+ * @param [in] dev - The device structure.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t cli_parse(struct cli_desc *dev)
+{
+	uint8_t c = 0, rdy = 1;
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/cli.c]: should be rdy = 0

File: projects/ADuCM3029_demo_cn0435/src/cli.c [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+		case _CR:
+			uart_cmd = UART_TRUE;
+			break;
+		case _BS:
+			if(uart_line_index == 0)
+				break;
+			uart_line_index--;
+			break;
+		case _TB:
+			usr_uart_write_string(dev->uart_device, (uint8_t*)"\n>");
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/cli.c]: i don't think the cast is needed.

File: projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c [hunk 5]
```diff
@@ -0,0 +1,1034 @@
...
+
+	/* Allocate memory for descriptor */
+	dev = calloc(1, sizeof *dev);
+	if (!dev)
+		return -1;
+
+	/* Initialize descriptor */
+	ret = modbus_setup_get_slave_addr(dev, init_param, &dev->slave_id);
+	if(ret < 0)
+		return ret;
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c]: free dev ? (free willy ) :D

File: projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c [hunk 6]
```diff
@@ -0,0 +1,1034 @@
...
+ */
+static void modbus_get_function(struct modbus_slave *dev, uint8_t *req,
+				uint8_t *func_index)
+{
+	uint8_t function_number = sizeof(commands_array) /
+				  sizeof(typeof(*commands_array));
+	uint8_t i;
+
+	for(i = 0; i < function_number; i++) {
+		if(req[MODBUS_FUNCTION_PLACE] == commands_array[i]) {
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c]: no need for brackets

File: projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c [hunk 7]
```diff
@@ -0,0 +1,1034 @@
...
+	}
+
+	switch(commands_array[i]) {
+	case READ_OUTPUT_HOLDING_REGISTERS:
+	case WRITE_SINGLE_OUTPUT_HOLDING_REGISTER:
+	case WRITE_MULTI_OUTPUT_HOLDING_REGISTER:
+		if(dev->out_holding_regs == NULL) {
+			dev->exception = MODBUS_EXCEPTION_ILLEGAL_FUNCTION;
+			dev->state = MBSLAVE_REPLY;
+			return;
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c]: i don't see the need for this return

File: projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c [hunk 8]
```diff
@@ -0,0 +1,1034 @@
...
+	switch(commands_array[i]) {
+	case READ_DISCRETE_OUTPUT_COIL:
+	case WRITE_SINGLE_DISCRETE_OUTPUT_COIL:
+	case WRITE_MULTI_DISCRETE_OUTPUT_COILS:
+		strt_reg = modbus_get_reg(dev, start_range, DISCRETE_OUTPUT_COILS);
+		max_reg = modbus_get_reg(dev, max_range, DISCRETE_OUTPUT_COILS);
+		if(strt_reg == NULL || max_reg == NULL) {
+			dev->exception = MODBUS_EXCEPTION_ILLEGAL_DATA_ADDRESS;
+			dev->state = MBSLAVE_REPLY;
+			return;
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.c]: no need for return

File: projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.h [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+	uint16_t out_coils_range;
+	uint16_t in_contacts_range;
+	uint16_t in_reg_range;
+	uint16_t holding_reg_range;
+};
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+//TODO: COMMENT modbus_setup
```
Comment (mirceacaprioru) [inline projects/ADuCM3029_demo_cn0435/src/mb_slave_data_link.h]: remove TODOs ?

File: projects/ADuCM3029_demo_cn0435/src/cli.c (hunk 1 above)
Comment (adrimbarean) [reply projects/ADuCM3029_demo_cn0435/src/cli.c]: No

**Comment by mirceacaprioru** [comment]
Looks good overall, just some remarks.

**Comment by adrimbarean** [comment]
V1 -> V2:
 - initialize cli.c:rdy with 0 and test for rdy instead of rdy == 1;
 - remove memory leak in mb_slave_data_link;
 - remove unnecessary brackets;
 - remove unnecessary returns from switch cases;
 - remove todos.

**Pull request merged**